compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEConvertToDma.cpp

```diff
@@ -146,129 +154,400 @@ LogicalResult unPackDmaInputs(IREE::LinalgExt::UnPackOp unPackOp,
+                                    SmallVector<OpFoldResult> &sizes,
+                                    SmallVector<OpFoldResult> &strides) {
+  MLIRContext *ctx = expandShapeOp.getContext();
+  auto reassociationIndices = expandShapeOp.getReassociationIndices();
+  ArrayRef<int64_t> resultShape = expandShapeOp.getType().getShape();
+
+  // Sizes.
+  SmallVector<OpFoldResult> newSizes(resultShape.size());
+  for (int i = 0; i < resultShape.size(); i++) {
+    newSizes[i] = getAsIndexOpFoldResult(ctx, resultShape[i]);
+  }
+
+  // Strides.
+  SmallVector<OpFoldResult> newStrides(resultShape.size());
+  for (auto reassociation : llvm::enumerate(reassociationIndices)) {
+    auto index = reassociation.index();
+    auto dims = reassociation.value();
+    int64_t cum = getConstantIntValue(strides[index]).value();
+    for (uint64_t i = 0; i < dims.size(); i++) {
+      auto d = dims[dims.size() - i - 1];
+      newStrides[d] = getAsIndexOpFoldResult(ctx, cum);
+      cum *= resultShape[d];
+    }
+  }
+
+  // Offsets. For now we don't do any arithmetic to split the offset across
+  // dimensions, in theory we need to split the offset amongst the reassociation
+  // indices, but for now I'm just putting the offset on the inner most
+  // dimension.
+  SmallVector<OpFoldResult> newOffsets(resultShape.size());
+  for (int i = 0; i < resultShape.size(); i++) {
+    newOffsets[i] = getAsIndexOpFoldResult(ctx, 0);
+  }
+  for (auto reassociation : llvm::enumerate(reassociationIndices)) {
+    auto index = reassociation.index();
+    auto dims = reassociation.value();
+    newOffsets[dims.back()] = offsets[index];
+  }
+
+  sizes = newSizes;
+  offsets = newOffsets;
+  strides = newStrides;
+  return success();
+}
 
-    assert(offsets.size() == sizes.size() && sizes.size() == strides.size() &&
-           "mismatch in the number of offsets, sizes and strides");
-
-    // Handle the case where some dimensions are dropped in the subview:
-    llvm::SmallBitVector droppedDims = subviewOp.getDroppedDims();
-    uint64_t insertionIndex{0};
-    for (uint64_t extractionIndex = 0; extractionIndex < offsets.size();
-         ++extractionIndex) {
-      if (!droppedDims[extractionIndex]) {
-        offsets[insertionIndex] = offsets[extractionIndex];
-        sizes[insertionIndex] = sizes[extractionIndex];
-        strides[insertionIndex] = strides[extractionIndex];
-        insertionIndex++;
+
+/// Update the offsets, sizes, and strides from a subview operation.
+LogicalResult updateFromSubView(memref::SubViewOp subviewOp,
+                                SmallVector<OpFoldResult> &offsets,
+                                SmallVector<OpFoldResult> &sizes,
+                                SmallVector<OpFoldResult> &strides) {
+  assert(offsets.size() == subviewOp.getMixedSizes().size());
+
+  OpBuilder builder(subviewOp.getContext());
+  builder.setInsertionPoint(subviewOp);
+  offsets = getIndexOpFoldResultSum(builder, subviewOp.getLoc(), offsets,
```

Comment:
It's better to add some comments here or above the utility function of how offsets are generated.